### PR TITLE
feat: discard batch safely after commit batch

### PR DESCRIPTION
**Pull request by rfyiamcool**
## change

1. discard batch safely.
2. lock/unlock safely.

#### refer:

- [https://pkg.go.dev/github.com/dgraph-io/badger#Txn.Discard](https://pkg.go.dev/github.com/dgraph-io/badger#Txn.Discard)
- [https://pkg.go.dev/github.com/dgraph-io/badger#WriteBatch.Cancel](https://pkg.go.dev/github.com/dgraph-io/badger#WriteBatch.Cancel)

### CODE REVIEW

**Comment by rfyiamcool** [comment]
add ut for discard.

**Comment by roseduan** [comment]
I think this is a little complicated, like we have to maintain the `locked` field.

A better simple solution is to add a `Rollback` function:
* clear pendingWrites if any
* make `rollbacked` field as true
* unlock
* add check `if rollbacked`, just return an error

so the user will either call commit or rollback.

**Comment by rfyiamcool** [comment]
> 

done, review.

**Comment by rfyiamcool** [comment]
> I think this is a little complicated, like we have to maintain the `locked` field.
> 
> A better simple solution is to add a `Rollback` function:
> 
> * clear pendingWrites if any
> * make `rollbacked` field as true
> * unlock
> * add check `if rollbacked`, just return an error
> 
> so the user will either call commit or rollback.

done, review.

**Comment by roseduan** [comment]
I think the `locked` field can be removed?

File: examples/batch/main.go [hunk 1]
```diff
@@ -20,6 +22,7 @@ func main() {
 
 	// create a batch
 	batch := db.NewBatch(rosedb.DefaultBatchOptions)
+	defer batch.Rollback()
```
Comment (roseduan) [inline examples/batch/main.go:25]: use will either call `Commit` or `Rollback`, so need to call this in defer

Comment (rfyiamcool) [reply examples/batch/main.go]: done

**Pull request merged**